File: protocol/impls/py/secp256k1.py
"""§4 Strategy B — real secp256k1 (self-rolled, stdlib-only).

A genuine elliptic-curve oracle to replace the injected curve pseudo-funcs of the
§4 attribution shell: field arithmetic mod p = 2^256 − 2^32 − 977, point ops, ECDSA
verify, and RFC-6979 deterministic signing (HMAC-SHA256 nonce, low-S normalized).
NOT constant time — a verifier/test oracle, never touches production secret keys.
Mirrors impls/c/src/secp256k1.c byte-for-byte (the C impl is normative). Python's
native int is the bignum; HMAC via stdlib hmac/hashlib.

Pinned in SPEC-conformance.md §13 (curve-vector set) + SPEC-RATIONALE.md §11.
"""
import hashlib
import hmac as _hmac

# p = 2^256 − 2^32 − 977 ; n = group order ; G = generator. Big-endian values.
SECP_P = 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEFFFFFC2F
SECP_N = 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03BBFD25E8CD0364141
SECP_N_HALF = SECP_N >> 1
SECP_GX = 0x79BE667EF9DCBBAC55A06295CE870B07029BFCDB2DCE28D959F2815B16F81798
SECP_GY = 0x483ADA7726A3C4655DA4FBFC0E1108A8FD17B448A68554199C47D08FFB10D4B8


def _sha256(d):
    return hashlib.sha256(d).digest()


def _hmac_sha256(key, msg):
    return _hmac.new(key, msg, hashlib.sha256).digest()


# ── field arithmetic mod p (Python int does the bignum) ───────────────────────
def _inv_p(a):
    return pow(a % SECP_P, SECP_P - 2, SECP_P)


def _sqrt_p(a):
    # p ≡ 3 (mod 4): sqrt = a^((p+1)/4)
    return pow(a % SECP_P, (SECP_P + 1) >> 2, SECP_P)


# ── points (affine over GF(p); None == point at infinity) ─────────────────────
def _point_add(p, q):
    if p is None:
        return q
    if q is None:
        return p
    x1, y1 = p
    x2, y2 = q
    if x1 == x2:
        if (y1 + y2) % SECP_P == 0:
            return None              # P + (−P)
        return _point_double(p)      # P == Q
    lam = ((y2 - y1) * _inv_p(x2 - x1)) % SECP_P
    x3 = (lam * lam - x1 - x2) % SECP_P
    y3 = (lam * (x1 - x3) - y1) % SECP_P
    return (x3, y3)


def _point_double(p):
    if p is None:
        return None
    x1, y1 = p
    if y1 % SECP_P == 0:
        return None
    lam = ((3 * x1 * x1) * _inv_p(2 * y1)) % SECP_P
    x3 = (lam * lam - 2 * x1) % SECP_P
    y3 = (lam * (x1 - x3) - y1) % SECP_P
    return (x3, y3)


def _scalar_mul(k, p):
    # double-and-add over the 256-bit big-endian scalar, MSB first
    r = None
    for bit in range(255, -1, -1):
        r = _point_double(r)
        if (k >> bit) & 1:
            r = _point_add(r, p)
    return r


_G = (SECP_GX, SECP_GY)


# ── pubkey decode + on-curve ──────────────────────────────────────────────────
def _pub_decode(pub):
    """Decode a 33-byte compressed (0x02/0x03) or 65-byte uncompressed (0x04)
    pubkey to (x, y). Returns None on bad length/prefix, X≥p, non-residue."""
    if len(pub) == 33 and pub[0] in (0x02, 0x03):
        x = int.from_bytes(pub[1:33], "big")
        if x >= SECP_P:
            return None
        rhs = (pow(x, 3, SECP_P) + 7) % SECP_P
        beta = _sqrt_p(rhs)
        if (beta * beta) % SECP_P != rhs:
            return None              # not a quadratic residue ⇒ off curve
        want_odd = (pub[0] == 0x03)
        if (beta & 1) != want_odd:
            beta = (SECP_P - beta) % SECP_P
        return (x, beta)
    if len(pub) == 65 and pub[0] == 0x04:
        x = int.from_bytes(pub[1:33], "big")
        y = int.from_bytes(pub[33:65], "big")
        if x >= SECP_P or y >= SECP_P:
            return None
        rhs = (pow(x, 3, SECP_P) + 7) % SECP_P
        if (y * y) % SECP_P != rhs:
            return None
        return (x, y)
    return None


def on_curve(pub):
    """1 iff the canonically-encoded pubkey decodes to a point on the curve."""
    return 1 if _pub_decode(pub) is not None else 0


# ── ECDSA verify (does NOT enforce low-S) ─────────────────────────────────────
def ecdsa_verify(hash32, r32, s32, pub):
    """r32/s32 = 32-byte big-endian scalar VALUES. 1 iff valid for (hash, pub)."""
    q = _pub_decode(pub)
    if q is None:
        return 0
    r = int.from_bytes(r32, "big")
    s = int.from_bytes(s32, "big")
    if r == 0 or r >= SECP_N:
        return 0
    if s == 0 or s >= SECP_N:
        return 0
    z = int.from_bytes(hash32, "big") % SECP_N
    w = pow(s, SECP_N - 2, SECP_N)
    u1 = (z * w) % SECP_N
    u2 = (r * w) % SECP_N
    a = _scalar_mul(u1, _G)
    b = _scalar_mul(u2, q)
    rp = _point_add(a, b)
    if rp is None:
        return 0
    return 1 if (rp[0] % SECP_N) == r else 0


# ── RFC-6979 deterministic nonce + ECDSA sign ─────────────────────────────────
def _rfc6979_k(priv32, hash32):
    h1 = hash32
    h1o = (int.from_bytes(h1, "big") % SECP_N).to_bytes(32, "big")  # bits2octets
    x = priv32                                                       # int2octets
    V = b"\x01" * 32
    K = b"\x00" * 32
    K = _hmac_sha256(K, V + b"\x00" + x + h1o)
    V = _hmac_sha256(K, V)
    K = _hmac_sha256(K, V + b"\x01" + x + h1o)
    V = _hmac_sha256(K, V)
    while True:
        V = _hmac_sha256(K, V)                                       # T = V (qlen 256)
        k = int.from_bytes(V, "big")
        if 1 <= k < SECP_N:
            return V
        K = _hmac_sha256(K, V + b"\x00")
        V = _hmac_sha256(K, V)


def ecdsa_sign(priv32, hash32):
    """RFC-6979 deterministic ECDSA sign, low-S. Returns (r32, s32) or None."""
    d = int.from_bytes(priv32, "big")
    if d == 0 or d >= SECP_N:
        return None
    z = int.from_bytes(hash32, "big") % SECP_N
    feed = hash32
    for _attempt in range(64):
        kb = _rfc6979_k(priv32, feed)
        k = int.from_bytes(kb, "big")
        rp = _scalar_mul(k, _G)
        if rp is None:
            feed = _sha256(kb)
            continue
        r = rp[0] % SECP_N
        if r == 0:
            feed = _sha256(kb)
            continue
        kinv = pow(k, SECP_N - 2, SECP_N)
        s = (kinv * (z + r * d)) % SECP_N
        if s == 0:
            feed = _sha256(kb)
            continue
        if s > SECP_N_HALF:
            s = SECP_N - s
        return (r.to_bytes(32, "big"), s.to_bytes(32, "big"))
    return None


def pubkey(priv32):
    """33-byte compressed pubkey from a private scalar, or None."""
    d = int.from_bytes(priv32, "big")
    if d == 0 or d >= SECP_N:
        return None
    p = _scalar_mul(d, _G)
    if p is None:
        return None
    x, y = p
    prefix = 0x03 if (y & 1) else 0x02
    return bytes([prefix]) + x.to_bytes(32, "big")


# ── ECMH (Elliptic Curve Multiset Hash) ───────────────────────────────────────
# An accumulator is a 33-byte compressed point (0x02/0x03 ‖ X-be); the all-zero
# sentinel (prefix 0x00) is the identity ∞. Mirrors secp256k1.c's ECMH block.
ECMH_H2C_TAG = b"ECMHh2c1"


def ecmh_identity():
    return b"\x00" * 33


def _ecmh_ser(p):
    if p is None:
        return b"\x00" * 33
    x, y = p
    return bytes([0x03 if (y & 1) else 0x02]) + x.to_bytes(32, "big")


def _ecmh_load(in33):
    if in33[0] == 0:
        return None
    return _pub_decode(in33)


def ecmh_hash(pre):
    """Try-and-increment hash-to-curve. Returns (pt33, ctr); pt prefix ALWAYS 0x02."""
    ctr = 0
    while True:
        cb = bytes([ctr & 0xFF, (ctr >> 8) & 0xFF, (ctr >> 16) & 0xFF, (ctr >> 24) & 0xFF])
        h = _sha256(ECMH_H2C_TAG + pre + cb)
        x = int.from_bytes(h, "big") % SECP_P
        rhs = (pow(x, 3, SECP_P) + 7) % SECP_P
        beta = _sqrt_p(rhs)
        if (beta * beta) % SECP_P == rhs:
            return (bytes([0x02]) + x.to_bytes(32, "big"), ctr)
        ctr += 1


def ecmh_negate(pt33):
    if pt33[0] == 0:
        return pt33
    return bytes([pt33[0] ^ 1]) + pt33[1:]


def ecmh_add(acc33, pt33):
    a = _ecmh_load(acc33)
    p = _ecmh_load(pt33)
    return _ecmh_ser(_point_add(a, p))


def ecmh_selftest():
    """0 on success. Mirrors secp_ecmh_selftest (commutativity/identity/inverse/round-trip)."""
    fail = 0
    idp = ecmh_identity()
    pa = ecmh_hash(b"alpha")[0]
    pb = ecmh_hash(b"beta")[0]
    acc1 = ecmh_add(ecmh_add(ecmh_identity(), pa), pb)
    acc2 = ecmh_add(ecmh_add(ecmh_identity(), pb), pa)
    if acc1 != acc2:                                   # commutativity
        fail += 1
    if ecmh_add(ecmh_identity(), pa) != pa:            # identity: ∞ + P == P
        fail += 1
    npa = ecmh_negate(pa)
    if ecmh_add(ecmh_add(ecmh_identity(), pa), npa) != idp:   # inverse: P + (−P) == ∞
        fail += 1
    acc1 = ecmh_add(ecmh_identity(), pa)
    acc2 = ecmh_add(ecmh_add(acc1, pb), ecmh_negate(pb))      # add-then-remove round-trip
    if acc1 != acc2:
        fail += 1
    return fail


# ── self-check: constants + 2G KAT + n·G=∞ + decompress + sign/verify ─────────
def selftest():
    """0 on success, nonzero (failure count) otherwise. Mirrors secp_selftest."""
    fail = 0
    # N_HALF = N>>1
    if SECP_N_HALF != (SECP_N >> 1):
        fail += 1
    # G on curve via uncompressed encoding
    g_unc = b"\x04" + SECP_GX.to_bytes(32, "big") + SECP_GY.to_bytes(32, "big")
    if not on_curve(g_unc):
        fail += 1
    # 2G known-answer
    G2X = 0xC6047F9441ED7D6D3045406E95C07CD85C778E4B8CEF3CA7ABAC09B95C709EE5
    G2Y = 0x1AE168FEA63DC339A3C58419466CEAEEF7F632653266D0E1236431A950CFE52A
    p2 = _scalar_mul(2, _G)
    if p2 is None or p2[0] != G2X or p2[1] != G2Y:
        fail += 1
    # n·G == ∞
    if _scalar_mul(SECP_N, _G) is not None:
        fail += 1
    # decompress round-trip: compress G (Gy even ⇒ 0x02), decode, compare Y
    gc = b"\x02" + SECP_GX.to_bytes(32, "big")
    dec = _pub_decode(gc)
    if dec is None or dec[1] != SECP_GY:
        fail += 1
    # sign / verify round-trip over deterministic keys + tamper check
    for t in range(1, 5):
        priv = (t * 7 + 1).to_bytes(32, "big")
        pub = pubkey(priv)
        if pub is None:
            fail += 1
            continue
        msg = bytes((i * 13 + t) & 0xFF for i in range(32))
        mh = _sha256(msg)
        sig = ecdsa_sign(priv, mh)
        if sig is None:
            fail += 1
            continue
        r, s = sig
        if not ecdsa_verify(mh, r, s, pub):
            fail += 1
        mh2 = bytes([mh[0] ^ 0x01]) + mh[1:]
        if ecdsa_verify(mh2, r, s, pub):
            fail += 1
    return fail


if __name__ == "__main__":
    import sys
    f = selftest()
    print("secp selftest", "ok" if f == 0 else "FAIL=%d" % f)
    sys.exit(0 if f == 0 else 1)

File: protocol/impls/ts/src/secp256k1.ts
// §4 Strategy B — real secp256k1 (clean-room TypeScript port of impls/c/src/secp256k1.c).
//
// Genuine elliptic-curve math over the field p = 2^256 − 2^32 − 977: field arithmetic,
// Jacobian point ops + double-and-add scalar multiply, ECDSA verify, and RFC-6979
// deterministic signing (HMAC-SHA256 nonce, low-S normalized). Native BigInt for all
// bignum (no limbs). NOT constant time — verifier/test oracle only, never secrets in prod.
// Mirrors C semantics exactly so the §13 curve-vector set is byte-identical cross-language.
import type { Bytes } from "./bytes.ts";
import { sha256 } from "./sha256.ts";

// ─── curve constants (ported from secp256k1.c; big-endian value semantics) ─────────────────────
export const P = 0xfffffffffffffffffffffffffffffffffffffffffffffffffffffffefffffc2fn; // 2^256−2^32−977
export const N = 0xfffffffffffffffffffffffffffffffebaaedce6af48a03bbfd25e8cd0364141n;
export const N_HALF = N >> 1n;
export const GX = 0x79be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798n;
export const GY = 0x483ada7726a3c4655da4fbfc0e1108a8fd17b448a68554199c47d08ffb10d4b8n;

const MASK256 = (1n << 256n) - 1n;

// ─── 32-byte big-endian <-> bigint ─────────────────────────────────────────────────────────────
export function be32(v: bigint): Bytes {
  const out = new Uint8Array(32);
  let x = v & MASK256;
  for (let i = 31; i >= 0; i--) { out[i] = Number(x & 0xffn); x >>= 8n; }
  return out;
}
function fromBE(b: Bytes, off = 0, len = 32): bigint {
  let v = 0n;
  for (let i = 0; i < len; i++) v = (v << 8n) | BigInt(b[off + i]);
  return v;
}

// ─── field arithmetic mod p ──────────────────────────────────────────────────────────────────
const mod = (a: bigint, m: bigint): bigint => { const r = a % m; return r < 0n ? r + m : r; };
const fadd = (a: bigint, b: bigint): bigint => mod(a + b, P);
const fsub = (a: bigint, b: bigint): bigint => mod(a - b, P);
const fmul = (a: bigint, b: bigint): bigint => mod(a * b, P);
const fsqr = (a: bigint): bigint => mod(a * a, P);

function fpow(a: bigint, e: bigint): bigint {
  let acc = 1n, base = mod(a, P), exp = e;
  while (exp > 0n) {
    if (exp & 1n) acc = fmul(acc, base);
    base = fsqr(base);
    exp >>= 1n;
  }
  return acc;
}
const finv = (a: bigint): bigint => fpow(a, P - 2n);            // a^(p-2)
const fsqrt = (a: bigint): bigint => fpow(a, (P + 1n) >> 2n);   // a^((p+1)/4); p≡3 mod 4

// ─── points (Jacobian: affine = (X/Z^2, Y/Z^3); Z==0 ⇒ infinity) ──────────────────────────────
type Jac = { X: bigint; Y: bigint; Z: bigint; inf: boolean };
const jacInf = (): Jac => ({ X: 1n, Y: 1n, Z: 0n, inf: true });
const jacFromAffine = (x: bigint, y: bigint): Jac => ({ X: x, Y: y, Z: 1n, inf: false });

function jacDouble(p: Jac): Jac {
  if (p.inf || p.Y === 0n) return jacInf();
  const A = fsqr(p.X);
  const B = fsqr(p.Y);
  const C = fsqr(B);
  let t = fsqr(fadd(p.X, B));
  t = fsub(fsub(t, A), C);
  const D = fadd(t, t);
  const E = fadd(fadd(A, A), A);                                // 3A
  const F = fsqr(E);
  const X3 = fsub(fsub(F, D), D);                              // F - 2D
  let y3 = fmul(E, fsub(D, X3));
  let t2 = fadd(fadd(fadd(C, C), fadd(C, C)), fadd(fadd(C, C), fadd(C, C))); // 8C
  const Y3 = fsub(y3, t2);
  const Z3 = fadd(fmul(p.Y, p.Z), fmul(p.Y, p.Z));            // 2·Y·Z
  return { X: X3, Y: Y3, Z: Z3, inf: false };
}
function jacAdd(p: Jac, q: Jac): Jac {
  if (p.inf) return { ...q };
  if (q.inf) return { ...p };
  const Z1Z1 = fsqr(p.Z), Z2Z2 = fsqr(q.Z);
  const U1 = fmul(p.X, Z2Z2), U2 = fmul(q.X, Z1Z1);
  const S1 = fmul(fmul(p.Y, q.Z), Z2Z2);                       // Y1·Z2^3
  const S2 = fmul(fmul(q.Y, p.Z), Z1Z1);                       // Y2·Z1^3
  if (U1 === U2) {
    if (S1 !== S2) return jacInf();                            // P + (−P)
    return jacDouble(p);                                        // P == Q
  }
  const H = fsub(U2, U1), R = fsub(S2, S1);
  const HH = fsqr(H), HHH = fmul(H, HH), V = fmul(U1, HH);
  const X3 = fsub(fsub(fsqr(R), HHH), fadd(V, V));            // R^2 - HHH - 2V
  const Y3 = fsub(fmul(R, fsub(V, X3)), fmul(S1, HHH));       // R·(V-X3) - S1·HHH
  const Z3 = fmul(fmul(p.Z, q.Z), H);                          // Z1·Z2·H
  return { X: X3, Y: Y3, Z: Z3, inf: false };
}
// R = scalar·P (double-and-add, MSB first over 256 bits) — mirrors C jac_mul byte loop.
function jacMul(p: Jac, k: bigint): Jac {
  let acc = jacInf();
  for (let bit = 255; bit >= 0; bit--) {
    acc = jacDouble(acc);
    if ((k >> BigInt(bit)) & 1n) acc = jacAdd(acc, p);
  }
  return acc;
}
function jacAffine(p: Jac): { x: bigint; y: bigint } | null {
  if (p.inf || p.Z === 0n) return null;
  const zinv = finv(p.Z);
  const zinv2 = fsqr(zinv);
  const zinv3 = fmul(zinv2, zinv);
  return { x: fmul(p.X, zinv2), y: fmul(p.Y, zinv3) };
}
const secpG = (): Jac => jacFromAffine(GX, GY);

// ─── pubkey decode + on-curve ──────────────────────────────────────────────────────────────────
function rhsCurve(x: bigint): bigint { return fadd(fmul(fsqr(x), x), 7n); } // x^3 + 7
// decode to affine; returns null off-curve / bad encoding (mirrors C pub_decode).
function pubDecode(pub: Bytes): { x: bigint; y: bigint } | null {
  const plen = pub.length;
  if (plen === 33 && (pub[0] === 0x02 || pub[0] === 0x03)) {
    const x = fromBE(pub, 1, 32);
    if (x >= P) return null;
    const rhs = rhsCurve(x);
    let beta = fsqrt(rhs);
    if (fsqr(beta) !== rhs) return null;                       // non-residue ⇒ off curve
    const wantOdd = pub[0] === 0x03;
    if ((beta & 1n) === 1n !== wantOdd) beta = fsub(P, beta);
    return { x, y: beta };
  }
  if (plen === 65 && pub[0] === 0x04) {
    const x = fromBE(pub, 1, 32);
    const y = fromBE(pub, 33, 32);
    if (x >= P || y >= P) return null;
    if (fsqr(y) !== rhsCurve(x)) return null;
    return { x, y };
  }
  return null;
}
export function onCurve(pub: Bytes): boolean {
  return pubDecode(pub) !== null;
}

// ─── ECDSA verify ────────────────────────────────────────────────────────────────────────────
export function ecdsaVerify(hash32: Bytes, r32: Bytes, s32: Bytes, pub: Bytes): boolean {
  const q = pubDecode(pub);
  if (!q) return false;
  const r = fromBE(r32), s = fromBE(s32);
  if (r === 0n || r >= N) return false;                        // 1 ≤ r < n
  if (s === 0n || s >= N) return false;                        // 1 ≤ s < n
  const z = mod(fromBE(hash32), N);                            // z mod n
  const w = modInv(s, N);                                       // s^{-1}
  const u1 = mod(z * w, N), u2 = mod(r * w, N);
  const Rj = jacAdd(jacMul(secpG(), u1), jacMul(jacFromAffine(q.x, q.y), u2));
  const aff = jacAffine(Rj);
  if (!aff) return false;                                       // R == ∞
  return mod(aff.x, N) === r;
}

// ─── scalar arithmetic mod n ───────────────────────────────────────────────────────────────────
function modInv(a: bigint, m: bigint): bigint {
  // extended Euclid; a in [1,m). Returns a^{-1} mod m (m prime / coprime).
  let [old_r, r] = [mod(a, m), m];
  let [old_s, s] = [1n, 0n];
  while (r !== 0n) {
    const qq = old_r / r;
    [old_r, r] = [r, old_r - qq * r];
    [old_s, s] = [s, old_s - qq * s];
  }
  return mod(old_s, m);
}

// ─── HMAC-SHA256 (on top of the suite's self-rolled sha256) ────────────────────────────────────
function hmacSha256(key: Bytes, msg: Bytes): Bytes {
  const block = 64;
  let k = key;
  if (k.length > block) k = sha256(k);
  const kpad = new Uint8Array(block);
  kpad.set(k, 0);
  const ki = new Uint8Array(block), ko = new Uint8Array(block);
  for (let i = 0; i < block; i++) { ki[i] = kpad[i] ^ 0x36; ko[i] = kpad[i] ^ 0x5c; }
  const inner = sha256(cat(ki, msg));
  return sha256(cat(ko, inner));
}
function cat(a: Bytes, b: Bytes): Bytes {
  const out = new Uint8Array(a.length + b.length);
  out.set(a, 0); out.set(b, a.length);
  return out;
}

// ─── RFC-6979 nonce + ECDSA sign ───────────────────────────────────────────────────────────────
// First valid k in [1,n) from the HMAC_DRBG stream keyed by (priv, h1). Mirrors C rfc6979_k.
function rfc6979K(priv32: Bytes, hash32: Bytes): bigint {
  const h1o = be32(mod(fromBE(hash32), N));                     // bits2octets(h1) = (h1 mod n) BE
  let V = new Uint8Array(32).fill(0x01);
  let K = new Uint8Array(32).fill(0x00);
  // K = HMAC_K(V ‖ 0x00 ‖ priv ‖ h1o)
  K = hmacSha256(K, cat(cat(V, Uint8Array.of(0x00)), cat(priv32, h1o)));
  V = hmacSha256(K, V);
  K = hmacSha256(K, cat(cat(V, Uint8Array.of(0x01)), cat(priv32, h1o)));
  V = hmacSha256(K, V);
  for (;;) {
    V = hmacSha256(K, V);                                       // T = V (qlen==256 ⇒ one block)
    const k = fromBE(V);
    if (k !== 0n && k < N) return k;
    K = hmacSha256(K, cat(V, Uint8Array.of(0x00)));
    V = hmacSha256(K, V);
  }
}
// RFC-6979 deterministic sign, low-S normalized. Returns {r,s} as 32-byte BE or null.
export function ecdsaSign(priv32: Bytes, hash32: Bytes): { r: Bytes; s: Bytes } | null {
  const d = fromBE(priv32);
  if (d === 0n || d >= N) return null;
  const z = mod(fromBE(hash32), N);
  let feed = Uint8Array.from(hash32);
  for (let attempt = 0; attempt < 64; attempt++) {
    const k = rfc6979K(priv32, feed);
    const Raff = jacAffine(jacMul(secpG(), k));
    if (!Raff) { feed = sha256(be32(k)); continue; }
    const r = mod(Raff.x, N);
    if (r === 0n) { feed = sha256(be32(k)); continue; }
    const kinv = modInv(k, N);
    let s = mod(kinv * mod(z + mod(r * d, N), N), N);
    if (s === 0n) { feed = sha256(be32(k)); continue; }
    if (s > N_HALF) s = N - s;                                  // low-S
    return { r: be32(r), s: be32(s) };
  }
  return null;
}
// derive the 33-byte compressed pubkey from a private scalar. Null if priv is 0 or ≥ n.
export function pubkey(priv32: Bytes): Bytes | null {
  const d = fromBE(priv32);
  if (d === 0n || d >= N) return null;
  const aff = jacAffine(jacMul(secpG(), d));
  if (!aff) return null;
  const out = new Uint8Array(33);
  out[0] = (aff.y & 1n) === 1n ? 0x03 : 0x02;
  out.set(be32(aff.x), 1);
  return out;
}

// ─── ECMH (Elliptic Curve Multiset Hash) ───────────────────────────────────────────────────────
// An accumulator is a 33-byte compressed point (0x02/0x03 ‖ X-be); the all-zero sentinel
// (prefix 0x00) is the identity ∞. Mirrors impls/c/src/secp256k1.c's ECMH block byte-for-byte.
const ECMH_H2C_TAG = Uint8Array.of(0x45, 0x43, 0x4d, 0x48, 0x68, 0x32, 0x63, 0x31); // "ECMHh2c1"

export function ecmhIdentity(): Bytes {
  return new Uint8Array(33);
}

// point → 33 bytes (∞ → zeros), real y-parity recompression.
function ecmhSer(p: Jac): Bytes {
  const aff = jacAffine(p);
  if (p.inf || !aff) return new Uint8Array(33);
  const out = new Uint8Array(33);
  out[0] = (aff.y & 1n) === 1n ? 0x03 : 0x02;
  out.set(be32(aff.x), 1);
  return out;
}
// 33 bytes → point (prefix 0x00 ⇒ ∞).
function ecmhLoad(in33: Bytes): Jac {
  if (in33[0] === 0) return jacInf();
  const dec = pubDecode(in33);
  if (!dec) return jacInf();
  return jacFromAffine(dec.x, dec.y);
}

// hash-to-curve, try-and-increment. Returns [33-byte even-Y compressed point, ctr used].
export function ecmhHash(pre: Bytes): { pt: Bytes; ctr: number } {
  for (let ctr = 0; ; ctr++) {
    // h = SHA256("ECMHh2c1" ‖ pre ‖ uint32_LE(ctr))
    const cb = Uint8Array.of(ctr & 0xff, (ctr >>> 8) & 0xff, (ctr >>> 16) & 0xff, (ctr >>> 24) & 0xff);
    const h = sha256(cat(cat(ECMH_H2C_TAG, pre), cb));
    const x = mod(fromBE(h), P);                  // x = SHA256(...) mod p
    const rhs = rhsCurve(x);                       // x³ + 7
    const beta = fsqrt(rhs);
    if (fsqr(beta) !== rhs) continue;              // not a QR ⇒ bump ctr
    const pt = new Uint8Array(33);
    pt[0] = 0x02;                                  // canonical even-Y
    pt.set(be32(x), 1);
    return { pt, ctr };
  }
}

export function ecmhNegate(pt33: Bytes): Bytes {
  const out = Uint8Array.from(pt33);
  if (out[0]) out[0] ^= 1;                         // identity (0x00) unchanged
  return out;
}

export function ecmhAdd(acc33: Bytes, pt33: Bytes): Bytes {
  const r = jacAdd(ecmhLoad(acc33), ecmhLoad(pt33));
  return ecmhSer(r);
}

// ─── self-check: pinned constants + 2G + n·G=∞ + decompress + sign/verify round-trip ───────────
export function secpSelftest(): number {
  let fail = 0;
  const F = (what: string): void => { fail++; if (process.env.SECP_KAT_VERBOSE) console.error("secp KAT FAIL:", what); };
  // constants: N_HALF = N>>1
  if ((N >> 1n) !== N_HALF) F("N_HALF");
  // G on curve (uncompressed)
  { const g = new Uint8Array(65); g[0] = 0x04; g.set(be32(GX), 1); g.set(be32(GY), 33);
    if (!onCurve(g)) F("G on curve"); }
  // 2G known-answer
  { const aff = jacAffine(jacMul(secpG(), 2n));
    const G2X = 0xc6047f9441ed7d6d3045406e95c07cd85c778e4b8cef3ca7abac09b95c709ee5n;
    const G2Y = 0x1ae168fea63dc339a3c58419466ceaeef7f632653266d0e1236431a950cfe52an;
    if (!aff || aff.x !== G2X || aff.y !== G2Y) F("2G value"); }
  // n·G == ∞
  { if (jacAffine(jacMul(secpG(), N)) !== null) F("n·G != inf"); }
  // decompress round-trip: compress G (even Y), decode, compare
  { const gc = new Uint8Array(33); gc[0] = 0x02; gc.set(be32(GX), 1);
    const dec = pubDecode(gc);
    if (!dec || dec.y !== GY) F("decompress G"); }
  // sign/verify round-trip + tamper over deterministic keys
  for (let t = 1; t <= 4; t++) {
    const priv = new Uint8Array(32); priv[31] = (t * 7 + 1) & 0xff;
    const pub = pubkey(priv);
    if (!pub) { F("pubkey"); continue; }
    const msg = new Uint8Array(32); for (let i = 0; i < 32; i++) msg[i] = (i * 13 + t) & 0xff;
    const mh = sha256(msg);
    const sig = ecdsaSign(priv, mh);
    if (!sig) { F("sign"); continue; }
    if (!ecdsaVerify(mh, sig.r, sig.s, pub)) F("verify");
    const mh2 = Uint8Array.from(mh); mh2[0] ^= 0x01;
    if (ecdsaVerify(mh2, sig.r, sig.s, pub)) F("tamper verify");
  }
  return fail;
}
